src/mesh/meshblock.hpp

```diff
@@ -65,9 +66,22 @@ KOKKOS_INLINE_FUNCTION void par_for_inner(const team_mbr_t &team_member, const i
   parthenon::par_for_inner(DEFAULT_INNER_LOOP_PATTERN, team_member, il, iu, function);
 }
 
+//! \class MeshBlockDevice
+//! \brief "device" view of MeshBlock for use in Kokkos contexts
+class MeshBlockDevice {
+ public:
+  // Inner loop default pattern
+  template <typename Function>
+  KOKKOS_INLINE_FUNCTION void par_for_inner(const team_mbr_t &team_member, const int &il,
+                                            const int &iu, const Function &function) const
+      noexcept {
+    parthenon::par_for_inner(DEFAULT_INNER_LOOP_PATTERN, team_member, il, iu, function);
+  }
+};
```

Comment:
This seems weird to me. Why not just expose a meshblock-independent `par_for_inner`? As this `MeshBlockDevice` has no internal state?

Reply:
I'm happy to. It just seemed deliberate that it was made a member function, so I wasn't sure if there was some longer term plan.

@pgrete ?

Reply:
The main reason it was a member function is that it uses `DEFAULT_INNER_LOOP_PATTERN` which is something we set at compile time and (some time ago) we decided to only provide abstractions in the `kokkos_abstraction` layer but nothing that changes behavior.

It's not entirely clear what we gain from using the MeshBlockDevice and think it's potentially confusing.
The inner loop never uses a "device" as that information is already carried over by the `team_member`.
So I'm in favor of keeping the original `pmb->par_for_inner` (as it is consistent with the use of a `pmb->par_for_outer` from my point of view).

Reply:
> So I'm in favor of keeping the original pmb->par_for_inner (as it is consistent with the use of a pmb->par_for_outer from my point of view).

That was deliberately removed - let me show you the warning we get when you use `pmb->par_for_inner` in `advection_package.cpp`:

```
../example/advection/advection_package.cpp(323): warning: calling a __host__ function from a __host__ __device__ function is not allowed

../example/advection/advection_package.cpp(326): warning: calling a __host__ function from a __host__ __device__ function is not allowed

../example/advection/advection_package.cpp(355): warning: calling a __host__ function from a __host__ __device__ function is not allowed

../example/advection/advection_package.cpp(359): warning: calling a __host__ function from a __host__ __device__ function is not allowed

../example/advection/advection_package.cpp(390): warning: calling a __host__ function from a __host__ __device__ function is not allowed

../example/advection/advection_package.cpp(394): warning: calling a __host__ function from a __host__ __device__ function is not allowed

/usr/local/cuda-11.0/bin/../targets/ppc64le-linux/include/vector_types.h(421): warning: calling a __host__ function("std::__shared_count<( ::__gnu_cxx::_Lock_policy)2> ::~__shared_count") from a __host__ __device__ function("") is not allowed

/usr/local/cuda-11.0/bin/../targets/ppc64le-linux/include/vector_types.h(421): warning: calling a __host__ function("std::__shared_count<( ::__gnu_cxx::_Lock_policy)2> ::~__shared_count") from a __host__ __device__ function("") is not allowed

/usr/local/cuda-11.0/bin/../targets/ppc64le-linux/include/vector_types.h(421): warning: calling a __host__ function("std::__shared_count<( ::__gnu_cxx::_Lock_policy)2> ::~__shared_count") from a __host__ __device__ function("") is not allowed

../example/advection/advection_package.cpp(330): warning: calling a __host__ function("std::__shared_count<( ::__gnu_cxx::_Lock_policy)2> ::__shared_count") from a __host__ __device__ function("") is not allowed

../example/advection/advection_package.cpp(330): warning: calling a __host__ function("std::__shared_count<( ::__gnu_cxx::_Lock_policy)2> ::~__shared_count") from a __host__ __device__ function("") is not allowed

../example/advection/advection_package.cpp(364): warning: calling a __host__ function("std::__shared_count<( ::__gnu_cxx::_Lock_policy)2> ::__shared_count") from a __host__ __device__ function("") is not allowed

../example/advection/advection_package.cpp(364): warning: calling a __host__ function("std::__shared_count<( ::__gnu_cxx::_Lock_policy)2> ::~__shared_count") from a __host__ __device__ function("") is not allowed

../example/advection/advection_package.cpp(399): warning: calling a __host__ function("std::__shared_count<( ::__gnu_cxx::_Lock_policy)2> ::__shared_count") from a __host__ __device__ function("") is not allowed

../example/advection/advection_package.cpp(399): warning: calling a __host__ function("std::__shared_count<( ::__gnu_cxx::_Lock_policy)2> ::~__shared_count") from a __host__ __device__ function("") is not allowed

../external/Kokkos/core/src/Kokkos_View.hpp(979): warning: calling a __host__ function("std::__shared_count<( ::__gnu_cxx::_Lock_policy)2> ::~__shared_count") from a __host__ __device__ function("") is not allowed

../external/Kokkos/core/src/Kokkos_View.hpp(979): warning: calling a __host__ function("std::__shared_count<( ::__gnu_cxx::_Lock_policy)2> ::~__shared_count") from a __host__ __device__ function("") is not allowed

../external/Kokkos/core/src/Kokkos_View.hpp(979): warning: calling a __host__ function("std::__shared_count<( ::__gnu_cxx::_Lock_policy)2> ::~__shared_count") from a __host__ __device__ function("") is not allowed
```

It's not 100% clear to me _why_ this is resulting in a "call" to the `__shared_count` destructor (I suspect it must be optimized out in practice, even if it technically exists). But I will say there's something eye-brow raising about accessing a `std::shared_ptr<MeshBlock>` in a device kernel, since 1) `std::shared_ptr` is not really GPU compatible and 2) `MeshBlock` is largely filled with member variables that aren't accessible from within the device kernel anyway. So that was the thought behind `MeshBlockData` - it was like, well, it seems like people want to be able to refer to the `MeshBlock` inside the device kernel, so this gives people a way to do it. But I'm open to just getting rid of the member function altogether.

Reply:
Yeah, I tried to scour the ptx, and, as far as I can tell, no call to `std::__shared_count<( ::__gnu_cxx::_Lock_policy)2> ::__shared_count` or `std::__shared_count<( ::__gnu_cxx::_Lock_policy)2> ::~__shared_count` are ever made.

Reply:
I think my preferred implmentation here would be to put `par_for_inner` in the `meshblock.hpp` header, but remove it from the meshblock class.

@pgrete , @AndrewGaspar would that satisfy constraints?

Reply:
sounds good to me.